nltk/parse/corenlp.py

```diff
@@ -0,0 +1,647 @@
+
+from nltk.parse.api import ParserI
+from nltk.tokenize.api import TokenizerI
+from nltk.parse.dependencygraph import DependencyGraph
+from nltk.tree import Tree
+
+_stanford_url = 'http://stanfordnlp.github.io/CoreNLP/'
+
+
+class CoreNLPServerError(EnvironmentError):
+    """Exceptions associated with the Core NLP server."""
+
+
+def try_port(port=0):
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.bind(('', port))
+
+    p = sock.getsockname()[1]
+    sock.close()
+
+    return p
+
+
+class CoreNLPServer(object):
+
+    _MODEL_JAR_PATTERN = r'stanford-corenlp-(\d+)\.(\d+)\.(\d+)-models\.jar'
+    _JAR = r'stanford-corenlp-(\d+)\.(\d+)\.(\d+)\.jar'
+
+    def __init__(
+        self, path_to_jar=None, path_to_models_jar=None, verbose=False,
+        java_options=None, corenlp_options=None, port=None,
+    ):
+        # find the most recent code and model jar
+        stanford_jar = max(
+            find_jar_iter(
+                self._JAR,
+                path_to_jar,
+                env_vars=('STANFORD_PARSER', 'STANFORD_CORENLP'),
+                searchpath=(),
+                url=_stanford_url,
+                verbose=verbose,
+                is_regex=True,
+            ),
+            key=lambda model_name: re.match(self._JAR, model_name)
+        )
+
+        if port is None:
+            try:
+                port = try_port(9000)
+            except socket.error:
+                port = try_port()
+
+            if corenlp_options:
+                corenlp_options.append(str(port))
+            else:
+                corenlp_options = [str(port)]
+        else:
+            try_port(port)
+
+        self.url = 'http://localhost:{}'.format(port)
+
+        model_jar = max(
+            find_jar_iter(
+                self._MODEL_JAR_PATTERN,
+                path_to_models_jar,
+                env_vars=('STANFORD_MODELS', 'STANFORD_CORENLP'),
+                searchpath=(),
+                url=_stanford_url,
+                verbose=verbose,
+                is_regex=True,
+            ),
+            key=lambda model_name: re.match(self._MODEL_JAR_PATTERN, model_name)
+        )
+
+        self.other_jars = tuple(find_jars_within_path(os.path.dirname(stanford_jar)))
+
+        self.verbose = verbose
+
+        self._classpath = (stanford_jar, model_jar) + self.other_jars
+
+        self.corenlp_options = corenlp_options or []
+        self.java_options = java_options or ['-mx1g']
+
+    def start(self):
+        cmd = ['edu.stanford.nlp.pipeline.StanfordCoreNLPServer']
+
+        if self.corenlp_options:
+            cmd.extend(self.corenlp_options)
+
+        # Configure java.
+        default_options = ' '.join(_java_options)
+        config_java(options=self.java_options, verbose=self.verbose)
+
+        try:
+            # TODO: it's probably a bad idea to pipe stdout, as it will
+            #       accumulate when lots of text is being parsed.
```

Comment:
`subprocess.DEVNULL` would be a great solution. I guess the main problem with `start()` that is should be a context manager, anything else will be hackish...

A compromise could be to use `subprocess.DEVNULL` if it's available.